src/lib.rs

```diff
@@ -344,11 +371,27 @@ pub struct Evaluator<'tcx> {
 
     /// The random number generator to use if Miri
     /// is running in non-deterministic mode
-    pub(crate) rng: Option<StdRng>
+    pub(crate) rng: Option<StdRng>,
+
+    /// Extra information needed for unwinding
+    /// We create this even in abort mode, so
+    /// that we can perform some basic validation
+    /// during panics
+    pub(crate) cached_data: CachedTypes<'tcx>,
+
+    /// Whether or not we are currently unwinding from
+    /// a panic
+    pub(crate) unwinding: bool,
```

Comment:
So we have *another* such flag here, on top of the one in `EvalContext` itself?

I still don't think either one is needed. The machine implicitly "knows" that it is unwinding by executing a block that leads to a `Resume` terminator (maybe indirectly). Any extra state trying to mirror that seems like a hack, and needs a very good justification.

Reply:
We definitely need to explicitly track the unwinding state. Since unwinding gets driven by the normal `run()` loop, we need a way to keep track of whether or not we should be popping off stack frames, or executing them as normal.

Reply:
I don't see why. We can tell this just by checking if we are executing a `Return` or `Resume` terminator.